analyzers/src/SonarAnalyzer.Utilities/RuleDetailBuilder.cs

```diff
@@ -33,177 +33,63 @@ namespace SonarAnalyzer.Utilities
-        }
-
-        private static IEnumerable<RuleDetail> GetRuleDetailFromRuleAttributes(Type analyzerType,
-            AnalyzerLanguage language)
-        {
-            return analyzerType.GetCustomAttributes<RuleAttribute>()
-                .Select(ruleAttribute => GetRuleDetail(ruleAttribute, analyzerType, language));
-        }
-
-        private static RuleDetail GetRuleDetail(RuleAttribute rule, Type analyzerType, AnalyzerLanguage language)
-        {
-            var resources = new ResourceManager("SonarAnalyzer.RspecStrings", analyzerType.Assembly);
-
-            var ruleDetail = new RuleDetail
+            var resources = language.LanguageName switch
             {
-                Key = rule.Key,
-                Type = GetBackwardsCompatibleType(resources.GetString($"{rule.Key}_Type")),
-                Title = resources.GetString($"{rule.Key}_Title"),
-                Severity = resources.GetString($"{rule.Key}_Severity"),
-                Status = resources.GetString($"{rule.Key}_Status"),
-                IsActivatedByDefault = bool.Parse(resources.GetString($"{rule.Key}_IsActivatedByDefault")),
-                Description = GetResourceHtml(rule, language),
-                Remediation = ToSonarQubeRemediationFunction(resources.GetString($"{rule.Key}_Remediation")),
-                RemediationCost = resources.GetString($"{rule.Key}_RemediationCost")
+                LanguageNames.CSharp => new ResourceManager("SonarAnalyzer.RspecStrings", typeof(Rules.CSharp.FlagsEnumZeroMember).Assembly),
+                LanguageNames.VisualBasic => new ResourceManager("SonarAnalyzer.RspecStrings", typeof(Rules.VisualBasic.FlagsEnumZeroMember).Assembly),
+                _ => throw new InvalidOperationException("Unexpected language: " + language)
             };
-
-            ruleDetail.Tags.AddRange(resources.GetString($"{rule.Key}_Tags").Split(new[] { ',' },
-                StringSplitOptions.RemoveEmptyEntries));
-
-            GetParameters(analyzerType, ruleDetail);
-            GetCodeFixNames(analyzerType, ruleDetail);
-
-            return ruleDetail;
+            return new RuleFinder()
+                .GetAnalyzerTypes(language)
+                .Select(x => (DiagnosticAnalyzer)Activator.CreateInstance(x))
+                .SelectMany(x => UniqueRuleIds(x).Select(id => new { Id = id, Type = x.GetType() }))
+                .GroupBy(x => x.Id)    // Same ruleId can be in multiple classes (see InvalidCastToInterface)
+                .Select(x => CreateRuleDetail(language, resources, x.Key, x.Select(item => item.Type)));
         }
 
-        private static Type GetCodeFixProviderType(Type analyzerType)
-        {
-            var typeName = analyzerType.FullName + CodeFixProviderSuffix;
-            return analyzerType.Assembly.GetType(typeName);
-        }
+        internal /*for testring*/ static IEnumerable<string> CodeFixTitles(Type codeFixType) =>
```

Comment:
The `/*for testring*/` is related to the `internal` keyword. Otherwise, it could be private.

Such thing is used in many places in the codebase:
https://github.com/SonarSource/sonar-dotnet/search?q=%22internal+%2F*+for+testing%22&type=code

I can remove it. (I hope to remove the whole class one day)

Reply:
I would remove this kind of comments. They can be confusing and easily get out of date. When you call a method, you shouldn't need to look if it has any comments about where it's used :).

Reply:
It was gone in `Review update 1`